### ISSUE TITLE
Enhancement: [strict-boolean-expressions] DisallowShortCircuit

### ISSUE THREAD

**Issue by maxballenger** [typescript-eslint/typescript-eslint#6657]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/strict-boolean-expressions/

### Description

I propose a new rule option for strict-boolean-expressions prohibiting the short circuit assignment pattern. This is somewhat commonly used but I have found it to be the source of many bugs. The classic example is using it for a default value of something which is optionally present. `undefined` is falsey, which will correctly lead to the default value being used, but 0 is also falsey, and an intentionally assigned 0 is usually not expected to trigger the default case.

### Fail

```typescript
// Using logical operator short-circuiting is allowed
const Component = () => {
  const entry = map.get('foo') || {};
  return entry && <p>Name: {entry.name}</p>;
};
```

### Pass

```typescript
// Using logical operator short-circuiting is not allowed
const Component = () => {
  let entry = map.get('foo');
  if (entry == undefined) entry = {};
  return entry && <p>Name: {entry.name}</p>;
};
```

### Additional Info

[these docs](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md) say it is allowed.

**Comment by bradzacher** [comment]
Have you got a different, more concrete example? Your example code makes little sense in the context of your request because the rule will report on this case due to the fact that the value is never false.

[playground](https://typescript-eslint.io/play#ts=5.0.1-rc&jsx=true&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3gWygBwC54BZfAHkxgEsUBzAGngG8UpsQjKbaBfAPgDcAKCHJ0WAMJJseVCBRYAvPAAUASniK+zIfESpM8eRhgBPTTnwA6WiAwqA5ADMkSBxoA+H5j2F64GACuMChGCmbwAGSRYSamVqzswr5CQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs+SAIZN6AD2LQUySnyaoMkRNGh9okcGAC+ILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A).

----

The rule checks short-circuit logical expressions - though it currently does not check the final expression of a short-circuit. This is on purpose because the general case for a short-circuit is to do a shorthand ternary expression - so the last expression is not used in a boolean context. [The rule will check the other expressions in the logical as they are being used in a boolean context.](https://typescript-eslint.io/play#ts=5.0.1-rc&jsx=true&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3gWygBwC54BZfAHkxgEsUBzAGnkptoD4BuAKE+XSwGEk2PKhAosAXngAKAJTwJreAG9O8RKkzwxGGAE8FOfADpaIDNIDkAMyRJL8gD6P4ly13VwMAVxgpt4vrwAGTBAbp6xhBitBgAFlwAvlxAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs+SAIZN6AD2LQUySnyaoMkRNGh9okcGAC+ILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A)

**Comment by bradzacher** [comment]
So the reason you don't get an error from your test function is because TS tried to be as narrow as possible and it makes the return type `0 | 1 | undefined`.

You then butt up against this bug https://github.com/typescript-eslint/typescript-eslint/issues/6624 because we currently just assume all literal types are truthy.

By manually annotating the return type, you're widening it to `number | undefined`, working around the above bug.

Again - this bug only happens if you return literal values AND you return a falsey literal value (eg `0` or `''`) AND the type unions in a nullish value.
